### PR TITLE
resourcedetectionprocessor: use opentelemetry-go for host id detection

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**

Use `openelemetry-go` to get `HostID`

This is supplementary for #18618 as result of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18618#pullrequestreview-1302045943

**Link to tracking Issue:** #18533

**Testing:** N/A

**Documentation:** N/A

### CODE REVIEW

**Comment by sumo-drosiek** [comment]
lint error seems to be unrelated: 
```
Check License finished successfully
/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/.tools/golangci-lint run --allow-parallel-runners --build-tags integration
level=error msg="Timeout exceeded: try increasing it by passing --timeout option"
make[2]: *** [../../Makefile.Common:166: lint] Error 4
make[2]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/otelcontribcol'
make[1]: *** [Makefile:167: cmd/otelcontribcol] Error 2
make[1]: *** Waiting for unfinished jobs....
level=error msg="Timeout exceeded: try increasing it by passing --timeout option"
make[2]: *** [../../Makefile.Common:166: lint] Error 4
make[2]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/cmd/configschema'
make[1]: *** [Makefile:167: cmd/configschema] Error 2
make[1]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib'
make: *** [Makefile:105: golint] Error 2
```

**Comment by astencel-sumo** [comment]
The way getting host ID is implemented in `gopsutil` vs the `github.com/panta/machineid` package is similar for Windows ([1](https://github.com/shirou/gopsutil/blob/v3.23.1/host/host_windows.go#L71), [2](https://github.com/panta/machineid/blob/v1.0.2/id_windows.go#L12)) and MacOS ([1](https://github.com/shirou/gopsutil/blob/v3.23.1/host/host_darwin.go#L25), [2](https://github.com/panta/machineid/blob/v1.0.2/id_darwin.go#L49)), but quite different for Linux (and Docker) ([1](https://github.com/shirou/gopsutil/blob/v3.23.1/host/host_linux.go#L38), [2](https://github.com/panta/machineid/blob/v1.0.2/id_linux.go#L38)). Just wanted to call this out. I'm not against using gopsutil, but I wonder if there are some corner cases that Dominik you might be aware of?

**Review by dmitryax — APPROVED** [review]
LGTM

**Comment by dmitryax** [comment]
@sumo-drosiek please rebase

**Comment by dmitryax** [comment]
@astencel-sumo, the difference itself is not a concern because the system detector `host.id` is not released yet. It was merged couple days ago https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18618. Do you think that `github.com/panta/machineid` is better in any way then gopsutil?

**Comment by sumo-drosiek** [comment]
> I'm not against using gopsutil, but I wonder if there are some corner cases that Dominik you might be aware of?

I'm not aware of any. Both `HostID` for `gopsutil` and `machineid` has started around the same time `2016, 2017` and gopsutil seems to be safer choice (actively developed). I wasn't aware of `HostID` and this is why I used `machineid` in the first place.

**Review by mx-psi — COMMENTED** [review]
The implementation used by the `machineID` dependency follows the note added to the spec on open-telemetry/opentelemetry-specification/pull/2978, while it's not clear to me that gopsutil follows it. I think that is a strong reason to **not** use gopsutil here, unless it turns out it's getting the same ID with a different method.

@sumo-drosiek @dmitryax do you know if the gopsutil id is different from the `machineid` id, or is it just a difference in the implementation?

**Comment by sumo-drosiek** [comment]
> The implementation used by the `machineID` dependency follows the note added to the spec on [open-telemetry/opentelemetry-specification/pull/2978](https://github.com/open-telemetry/opentelemetry-specification/pull/2978), while it's not clear to me that gopsutil follows it. I think that is a strong reason to **not** use gopsutil here, unless it turns out it's getting the same ID with a different method.
> 
> @sumo-drosiek @dmitryax do you know if the gopsutil id is different from the `machineid` id, or is it just a difference in the implementation?

I saw that description, but it says `may` instead of `should`.

They are different:

```
➜  projects sudo cat /etc/machine-id                
11112222333344441111222233334444
➜  projects sudo cat /sys/class/dmi/id/product_uuid 
77778888-7777-8888-7777-888877778888
➜  projects sudo cat /proc/sys/kernel/random/boot_id
66665555-6666-5555-6666-555566665555
```

**Comment by mx-psi** [comment]
I agree that either the implementation from your previous PR or the one on this PR would be compliant with the spec, but I still think we should follow the spec recommendation unless there is a good reason not to, for consistency with other detectors on language libraries and elsewhere. Is avoiding using the `machineid` dependency a strong enough reason to deviate from the spec recommendation?

**Comment by sumo-drosiek** [comment]
> Is avoiding using the machineid dependency a strong enough reason to deviate from the spec recommendation?

I wouldn't say that, I would consider it as implementation detail

**Comment by astencel-sumo** [comment]
> @astencel-sumo, the difference itself is not a concern because the system detector `host.id` is not released yet. It was merged couple days ago #18618.

Yes that's correct, I'm not worried about this.

> Do you think that `github.com/panta/machineid` is better in any way than gopsutil?

I think `gopsutil` is kind of a de-facto standard and I would much rather use this than the less popular option `panta/machineid`. But I have one big reservation about the implementation that `gopsutil` provides: it seems that the `HostID()` call returns a different value depending on whether the process is running as root vs. when it's not? When running as root, the value from `/sys/class/dmi/id/product_uuid` is returned, which is different from `/etc/machine-id`, which is what is being returned when running as non-root. This sounds to me like a very bad idea.

**Comment by Aneurysm9** [comment]
> When running as root, the value from `/sys/class/dmi/id/product_uuid` is returned, which is different from `/etc/machine-id`, which is what is being returned when running as non-root. This sounds to me like a very bad idea.

Agreed that this seems like a Bad Idea™.  Does the other alternative perform consistently regardless of permission level?

**Comment by dmitryax** [comment]
@astencel-sumo, I agree. That's odd behavior. It's worth filing an issue in gopsutil. Do you have a chance to follow up?

**Comment by astencel-sumo** [comment]
Created the issue here: https://github.com/shirou/gopsutil/issues/1420. Please let me know if I got something wrong (or just comment on the issue).

**Comment by mwear** [comment]
I have been doing similar work for collecting `host.id` in [the specification](https://github.com/open-telemetry/opentelemetry-specification/pull/3173), the [JS SDK](https://github.com/open-telemetry/opentelemetry-js/pull/3575), and [Go SDK](https://github.com/open-telemetry/opentelemetry-go/pull/3812). Supposing the spec work and the implementation for the Go SDK are on the right path, we could find a way to share this code and not have to take a dependency on an external library. Please comment on the spec and proposed go implementation if you have any opinions.

**Comment by dmitryax** [comment]
@sumo-drosiek Thanks for submitting the issue. Looks like we can get the desired functionality added in gopsutil

**Comment by MovieStoreGuy** [comment]
@sumo-drosiek , are you still working on this PR?

**Comment by perk-sumo** [comment]
Hey @MovieStoreGuy yes, we are going to work on this PR in the near future :)

**Comment by MovieStoreGuy** [comment]
Feel free to ping me once you're ready for a review.

**Comment by sumo-drosiek** [comment]
> @sumo-drosiek , are you still working on this PR?

I can work on it, but I'm little bit confused what is the current status. Do we need to extract code from `opentelemetry-go`, or just move it around there and then use it in `opentelemetry-collector-contrib`?

**Comment by astencel-sumo** [comment]
It looks like the "go.opentelemetry.io/otel/sdk" module is already being imported and [used](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.76.3/receiver/simpleprometheusreceiver/examples/federation/prom-counter/main.go#L30) in the contrib repo, so I believe referencing it is an option?

File: internal/metadataproviders/system/metadata.go [hunk 1]
```diff
@@ -128,6 +129,22 @@ func (p systemMetadataProvider) reverseLookup(ipAddresses []string) (string, err
...
+	)
+
+	if err != nil {
+		return "", fmt.Errorf("failed to obtain host id: %w", err)
+	}
+
+	iter := res.Iter()
+
+	for iter.Next() {
+		if iter.Attribute().Key == "host.id" {
```
Comment (astencel-sumo) [inline internal/metadataproviders/system/metadata.go:144]:
```suggestion
		if iter.Attribute().Key == conventions.AttributeHostID {
```

File: internal/metadataproviders/system/metadata.go [hunk 2]
```diff
@@ -128,6 +129,22 @@ func (p systemMetadataProvider) reverseLookup(ipAddresses []string) (string, err
 	return "", fmt.Errorf("reverseLookup failed to convert IP addresses to name: %w", err)
 }
 
-func (p systemMetadataProvider) HostID() (string, error) {
-	return machineid.ID()
+func (p systemMetadataProvider) HostID(ctx context.Context) (string, error) {
+	res, err := resource.New(ctx,
```
Comment (astencel-sumo) [inline internal/metadataproviders/system/metadata.go:133]: It's a bit "funny" how you need create a resource that includes the "host.id" attribute here, only to "peel" the resource to get just the "host.id" attribute value, and in the end you put that value in an attribute of another resource. This is due to the structure of the contrib's `system.Provider` interface :shrug:

File: .chloggen/drosiek-host-id-2.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: resourcedetectionprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: use gopsutil for host id detection
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply .chloggen/drosiek-host-id-2.yaml:8]:
```suggestion
note: use openelemetry-go for host id detection
```

**Review by andrzej-stencel — APPROVED** [review]
Great, thank you so much @sumo-drosiek :clap: 

@mwear it would be great to have your eyes on it, as this code uses the "host.id" code from opentelemetry-go.

@MovieStoreGuy also wanted to be pinged on this, so here you go :slightly_smiling_face:

**Review by MovieStoreGuy — APPROVED** [review]

File: internal/metadataproviders/system/metadata.go [hunk 4]
```diff
@@ -15,14 +15,16 @@
...
+	"context"
 	"fmt"
 	"net"
 	"os"
 	"runtime"
 	"strings"
 
 	"github.com/Showmax/go-fqdn"
-	"github.com/panta/machineid"
+	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply internal/metadataproviders/system/metadata.go:26]: It's funny how many versions are being used across the repository. I'm using the most popular one :see_no_evil: 

```
      3         conventions "go.opentelemetry.io/collector/semconv/v1.13.0"
      1         conventions "go.opentelemetry.io/collector/semconv/v1.16.0"
      6         conventions "go.opentelemetry.io/collector/semconv/v1.18.0"
    161         conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
      7         conventions "go.opentelemetry.io/collector/semconv/v1.8.0"
     19         conventions "go.opentelemetry.io/collector/semconv/v1.9.0"
      1         conventions "go.opentelemetry.io/collector/semconv/v{{ .SemConvVersion }}"
      3         // https://github.com/open-telemetry/opentelemetry-collector/blob/main/model/semconv/opentelemetry.go.
      1 // https://go.opentelemetry.io/collector/blob/main/model/semconv/opentelemetry.go.
      3         semconv "go.opentelemetry.io/collector/semconv/v1.13.0"
      1         semconv "go.opentelemetry.io/collector/semconv/v1.16.0"
      1         semconv "go.opentelemetry.io/collector/semconv/v1.18.0"
      3         semconv "go.opentelemetry.io/collector/semconv/v1.6.1"
      1         semconv "go.opentelemetry.io/collector/semconv/v1.8.0"
      3         semconv "go.opentelemetry.io/collector/semconv/v1.9.0"
      5         semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
```

[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply internal/metadataproviders/system/metadata.go:26]:
```suggestion
	conventions "go.opentelemetry.io/collector/semconv/v1.18.0"
```

[the remark this answers is not in the document]
Comment (mx-psi) [reply internal/metadataproviders/system/metadata.go]: I filed #22095

File: internal/metadataproviders/system/metadata.go [hunk 5]
```diff
@@ -128,6 +130,22 @@ func (p systemMetadataProvider) reverseLookup(ipAddresses []string) (string, err
 	return "", fmt.Errorf("reverseLookup failed to convert IP addresses to name: %w", err)
 }
 
-func (p systemMetadataProvider) HostID() (string, error) {
-	return machineid.ID()
+func (p systemMetadataProvider) HostID(ctx context.Context) (string, error) {
```
Comment (codeboten) [inline internal/metadataproviders/system/metadata.go:133]: Is this method tested anywhere? i don't see any specific tests around these changes.

**Review by codeboten — COMMENTED** [review]
Pinging @mwear @mx-psi for a review on this

**Review by mwear — APPROVED** [review]
LGTM

File: internal/metadataproviders/system/metadata.go (hunk 5 above)
Comment (sumo-drosiek) [reply internal/metadataproviders/system/metadata.go:133]: Not sure how to test it properly as extracting `resource` to interface for me looks like unnecessary complication. I am open for suggestions

**Comment by dmitryax** [comment]
@sumo-drosiek, please rebase

File: .chloggen/drosiek-host-id-2.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: resourcedetectionprocessor
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: use openelemetry-go for host id detection
```
[the remark this answers is not in the document]
Comment (atoulme) [reply .chloggen/drosiek-host-id-2.yaml:8]: There is a typo still

[the remark this answers is not in the document]
Comment (dmitryax) [reply .chloggen/drosiek-host-id-2.yaml:8]: @sumo-drosiek 

```suggestion
note: use opentelemetry-go library for host.id detection
```

[the remark this answers is not in the document]
Comment (codeboten) [reply .chloggen/drosiek-host-id-2.yaml:8]: @dmitryax you'll need `"` around that

**Review by dmitryax — APPROVED** [review]
LGTM

**Comment by mx-psi** [comment]
needs a rebase, we can merge after that

**Pull request merged**